### PR TITLE
Fix selected issues reported by Codacy

### PR DESCRIPTION
**Pull request by matthewrmshin**
For most external commands, redirect the null device to STDIN to prevent them from inheriting the parent's STDIN.

Remove unused code. Remove usage of `[]` and `{}` in function signatures. Improve some error handling. Improve some print statements. Improve some overridden method interfaces.

Fix various style complaints. Fix various scoping issues. Fix various attribute issues. Fix various abstract class issues.

Simply interface to `remrun`. Fix unused argument issue. Use a function to wrap an object to simplify interface as well as getting the naming convention correct.

### CODE REVIEW

**Comment by matthewrmshin** [comment]
(Still a bit more to do according to a local run of `pylint`. I'll also look into doing pickle-to-json when #2423 is done. It is not as straightforward as I thought, because pickle is used to serialize named tuples.)

**Comment by matthewrmshin** [comment]
That's all for now. I'll do more in a separate PR.

File: lib/cylc/cycling/integer.py [hunk 1]
```diff
@@ -109,6 +109,8 @@
         (r"^%(reps_1)s//%(end)s$", 4)
     ]
 ]
+del regex, format_num
```
[the remark this answers is not in the document]
Comment (hjoliver) [reply lib/cylc/cycling/integer.py]: Interesting.

**Comment by hjoliver** [comment]
There's now an unused pickle import in `task_pool.py`.

**Pull request merged**